source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -171,11 +172,43 @@ FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data, b
     result = FilterDataStatus::StopIterationAndBuffer;
     break;
 
+  case ProcessingMode::STREAMED: {
+    switch (openStream()) {
+    case StreamOpenState::Error:
+      return FilterDataStatus::StopIterationNoBuffer;
+    case StreamOpenState::IgnoreError:
+      return FilterDataStatus::Continue;
+    case StreamOpenState::Ok:
+      // Fall through
+      break;
+    }
+
+    auto next_chunk = std::make_unique<QueuedChunk>();
+    // Clear the current chunk and save it on the queue while it's processed
+    next_chunk->data.move(data);
+    next_chunk->end_stream = end_stream;
+    // Send the chunk, and ensure that we have watermarked so that we don't overflow
+    // memory while waiting for responses.
+    state.requestWatermark();
```

Comment:
We set watermark on every single chunk?

Reply:
That call is idempotent (that it, it has a flag so that it only calls the corresponding method on the stream context if it is not already set.) We leave the watermark "high" as long as there are stream messages outstanding.

Reply:
I'm hoping that resolves this but let me know otherwise.

Reply:
This seems a bit excessive. Shouldn't we set the high watermark when we trip over an actual watermark based on number of outbound requests?

Reply:
We can keep track of how much data is outstanding and only raise the watermark when it's above a threshold. I'm not sure what the threshold should be -- should it be the buffer limit that the user configured, or should it be half that? This buffering is happening entirely within this filter, so if there's other buffering happening elsewhere in Envoy I don't want to make the problem too much worse.

Reply:
Yeah, there are two things here:
1. When we raise the watermark, we're potentially causing HTTP flow control event logic to be triggered. 
2. There should be some high/low watermark distinction to avoid oscillating on (1).

I think if we track the number/size of outgoing buffered resources and provide some basic control over the upper bound we can consult it here, doing something like low watermark = high watermark / 2.

Reply:
I changed it to only raise the watermark when the total size of the queue is greater than the buffer limit / 2 -- that reduces the amount of watermarking a lot.